src/domain/relay/entities/contract-function.entity.ts

```diff
@@ -0,0 +1,56 @@
+import { Abi, decodeFunctionData } from 'viem';
+import { Hex } from 'viem/src/types/misc';
+
+/**
+ * Argument mapper representation for a contract function.
+ *
+ * The generic parameter represents the expected type that should be returned
+ * by the mapper.
+ */
+export type ContractFunctionArgsMapper<B> = {
+  name: string;
+  mapper: (args: readonly unknown[]) => B;
+};
+
+/**
+ * Decodes the arguments of the {@link contractFunction} given the provided
+ * {@link abi}.
+ *
+ * @param contractFunction - the contract function used to decode the arguments
+ * @param abi - the abi of the contract use to check the function call
+ * @param data - the data representing the function call
+ */
+export function decode<B>(
+  contractFunction: ContractFunctionArgsMapper<B>,
+  abi: readonly unknown[] | Abi,
+  data: Hex,
+): B {
+  const { functionName, args } = decodeFunctionData({
+    abi,
+    data,
+  });
+
+  if (contractFunction.name != functionName || !args)
+    throw Error(`Provided data is not a call to ${contractFunction.name}`);
+
+  return contractFunction.mapper(args);
```

Comment:
Mapping the arguments for every contract method will make each `ContractHelper` very "boilerplatey".

I think generalising `decode` to accept the method name and throw/return the `args` accordingly would suffice. It would simplify `ContractHelper`s for those which we need to support multiple methods, e.g. the Safe.

Let's discuss this on a call when we get the chance.

Reply:
One motivation for this approach was that `viem` returns `unknown[]` as a type of the function parameters thus the callers need to know that index 0 represents the argument with a given "name".

The goal of the abstraction was to return a structured object for the arguments for contracts. But I do agree that mapping every contract call can be excessive.

Reply:
Considering the point you made about only requiring `execTransaction` we could admittedly leave it as is for now but it will not "translate" well with interface changes.

Type safety is a valid concern. It doesn't seem like `typechain` supports `viem`, should we want to automatically generate them. It's a tough call - let's discuss it later today.

Reply:
> Considering the point you made about only requiring `execTransaction` we could admittedly leave it as is for now but it will not "translate" well with interface changes.
> 
> Type safety is a valid concern. It doesn't seem like `typechain` supports `viem`, should we want to automatically generate them. It's a tough call - let's discuss it later today.

I would say that this is ok: if the interface changes, this should break. The way it is implemented right now makes it compatible with any contract that has `execTransaction` as a function name, `to` as the 1st argument, `value` as the 2nd argument and `data` as the 3rd argument – we ignore the remaining arguments and their position.

The current `safe-contract.helper.spec.ts` instantiates a `SafeContractHelper` so the contracts' SDK will be used in the tests and they will fail if the SDK starts returning a different contract.